pkg/aws/eni/node.go

```diff
@@ -446,6 +477,73 @@ func (n *Node) determineAllocationAction() (*allocatableResources, error) {
 		limits:     limits,
 		totalENIs:  len(n.enis),
 	}
+
+	// Validate that the node still requires addresses to be released, the
+	// request may have been resolved in the meantime.
+	if option.Config.AwsReleaseExcessIps && n.stats.excessIPs > 0 {
+		for key, e := range n.enis {
+			scopedLog.WithFields(logrus.Fields{
+				fieldEniID:     e.ID,
+				"needIndex":    n.resource.Spec.ENI.FirstInterfaceIndex,
+				"index":        e.Number,
+				"addressLimit": a.limits.IPv4,
+				"numAddresses": len(e.Addresses),
+			}).Debug("Considering ENI for IP release")
+
+			if e.Number < n.resource.Spec.ENI.FirstInterfaceIndex {
+				continue
+			}
+
+			// Find free IP number on this ENI
+			ipsOnENI := n.resource.Status.ENI.ENIs[e.ID].Addresses
+			freeIpsOnENI := []string{}
+			for _, ip := range ipsOnENI {
+				_, ipUsed := n.resource.Status.IPAM.Used[ip]
+				// exclude primary IPs
+				if !ipUsed && ip != e.IP {
+					freeIpsOnENI = append(freeIpsOnENI, ip)
+				}
+			}
+			freeOnENICount := len(freeIpsOnENI)
+
+			if freeOnENICount <= 0 {
+				continue
+			}
+
+			scopedLog.WithFields(logrus.Fields{
+				fieldEniID:       e.ID,
+				"excessIPs":      n.stats.excessIPs,
+				"freeOnENICount": freeOnENICount,
+			}).Debug("ENI has IPs free")
```

Comment:
Let's change the message to `ENI has unused IPs that can be released` to make it clear that these are candidates for release

Reply:
done